server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -112,42 +132,77 @@ private void maybeScheduleNextRun(TimeValue time) {
     }
 
     // Helper method that starts the monitoring without a delay.
-    private void maybeScheduleNow() {
+    // Visible for testing
+    void maybeScheduleNow() {
         maybeScheduleNextRun(TimeValue.ZERO);
     }
 
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
-        if (prerequisitesFulfilled == false) {
-            prerequisitesFulfilled = event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)
-                && HealthMetadata.getFromClusterState(event.state()) != null;
-            maybeScheduleNow();
+        prerequisitesFulfilled = event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)
+            && HealthMetadata.getFromClusterState(event.state()) != null
+            && HealthNode.findHealthNode(event.state()) != null;
+        if (prerequisitesFulfilled) {
+            DiscoveryNode previous = HealthNode.findHealthNode(event.previousState());
+            DiscoveryNode current = HealthNode.findHealthNode(event.state());
+            if (Objects.equals(previous, current) == false) {
+                // The new health node (probably) does not have any information yet, so the last
+                // reported health info gets reset to null.
+                lastReportedDiskHealthInfo.set(null);
+            }
         }
+        maybeScheduleNow();
     }
 
-    // Visible for testing
-    void monitorHealth() {
-        if (inProgress.compareAndSet(false, true)) {
+    private void monitorHealth() {
+        if (inProgress.compareAndSet(false, true) && prerequisitesFulfilled) {
             ClusterState clusterState = clusterService.state();
             HealthMetadata healthMetadata = HealthMetadata.getFromClusterState(clusterState);
-            assert healthMetadata != null : "health metadata should have been initialized.";
-            DiskHealthInfo previousHealth = this.lastReportedDiskHealthInfo;
+            if (healthMetadata == null) {
+                logger.debug("Couldn't retrieve health metadata.");
+                inProgress.set(false);
+                return;
+            }
+            DiskHealthInfo previousHealth = this.lastReportedDiskHealthInfo.get();
             DiskHealthInfo currentHealth = diskCheck.getHealth(healthMetadata, clusterState);
             if (currentHealth.equals(previousHealth) == false) {
-                logger.debug("Health status changed from {} to {}", previousHealth, currentHealth);
-                this.lastReportedDiskHealthInfo = currentHealth;
+                String nodeId = clusterService.localNode().getId();
+                ActionListener<AcknowledgedResponse> listener = ActionListener.wrap(response -> {
+                    // Update the last reported value if there was no change, that would mean that there
+                    // is a new health node that might have not received the value, so we should retry.
```

Comment:
`Update the last reported value if there was no change,` - no change in health status? We shouldn't in that case right? But we're in the `if (currentHealth.equals(previousHealth) == false) {` branch so we know for sure there was a health status change.

Apologies, I dont' really follow what  this comment is meant to convey?

Reply:
No worries, I see it's not very clear. This comment is referring to this line: `lastReportedDiskHealthInfo.compareAndSet(previousHealth, currentHealth);`. What I intend to do with this is to ensure that we only set the `lastReportedDiskHealthInfo` if we know that the current health node received it. 

- A health node change will set `lastReportedDiskHealthInfo` to `null`, capturing that the current health node hasn't received any info.
- If the health node changed happened right after the info was sent to the previous health node and before the `lastReportedDiskHealthInfo` was updated, then `lastReportedDiskHealthInfo.compareAndSet(previousHealth, currentHealth)` will not go through. At the next monitoring run it will update the current health node.

While writing this though, I see there is a blind spot in this. If the `previousHealth` is `null`, then this will not catch it.

I will think of another solution :/.

Reply:
I changed the code to check the `lastSeenHealthNode` before it updates the last reported health info. This way we ensure that that when we update `lastReportedDiskHealthInfo` we are certain that the current health node received it. If not it will try again on the next try. 

What do you think?